leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/senml/LwM2mNodeSenMLJsonEncoder.java

```diff
@@ -111,54 +108,55 @@ public void visit(LwM2mResource resource) {
+            // Using request path as base name, and record doesn't have name
+            lwM2mResourceToSenMLRecord(null, resource);
         }
 
-        private void lwM2mResourceToSenMLRecord(String resourcePath, Long timestamp, LwM2mResource resource) {
+        private void lwM2mResourceToSenMLRecord(String recordName, LwM2mResource resource) {
             // create resource element
             if (resource.isMultiInstances()) {
                 for (Entry<Integer, ?> entry : resource.getValues().entrySet()) {
-                    // compute resource instance path
-                    String resourceInstancePath;
-                    if (resourcePath == null || resourcePath.isEmpty()) {
-                        resourceInstancePath = Integer.toString(entry.getKey());
+                    // compute record name for resource instance
+                    String resourceInstanceRecordName;
+                    if (recordName == null || recordName.isEmpty()) {
+                        resourceInstanceRecordName = Integer.toString(entry.getKey());
                     } else {
-                        resourceInstancePath = resourcePath + "/" + entry.getKey();
+                        resourceInstanceRecordName = recordName + "/" + entry.getKey();
                     }
 
-                    addSenMLRecord(resourceInstancePath, timestamp, resource, entry.getValue());
+                    addSenMLRecord(resourceInstanceRecordName, resource, entry.getValue());
                 }
             } else {
-                addSenMLRecord(resourcePath, timestamp, resource, resource.getValue());
+                addSenMLRecord(recordName, resource, resource.getValue());
             }
         }
 
-        private void addSenMLRecord(String resourcePath, Long timestamp, LwM2mResource resource, Object value) {
+        private void addSenMLRecord(String recordName, LwM2mResource resource, Object value) {
             // get type for this resource
             ResourceModel rSpec = model.getResourceModel(objectId, resource.getId());
             Type expectedType = rSpec != null ? rSpec.type : resource.getType();
 
-            // Create resource element
+            // Create SenML record
             SenMLRecord record = new SenMLRecord();
-            
-            if (records.isEmpty() && requestPath.isObjectInstance()) {
-                record.setBaseName(requestPath.toString());
-            }
-            
-            if(requestPath.isResource()) {
-                record.setBaseName(requestPath.toString());
-            }
-            else if(requestPath.isObjectInstance()) {
-                record.setName(resourcePath);
-                record.setTime(timestamp);
+
+            // Compute baseName and name for SenML record
+            String bn = requestPath.toString();
+            String n = recordName == null ? "" : recordName;
+
+            if (records.isEmpty()) {
+                if (!requestPath.isResource()) {
```

Comment:
I thing the right test is something like `!n.isEmpty()`

Reply:
Not fully sure I got your point. Do you mean to enhance line 144 or combine 146 and 147? When encoding object instance, 'n' is resource ID.

Reply:
This is for fixing, the point above https://github.com/eclipse/leshan/pull/677#discussion_r273973353

But we should agree on this point first :)

Reply:
this is for replacing test line 147.